vendors/espressif/boards/esp32/ports/ble/nimble/iot_ble_hal_gatt_server.c

```diff
@@ -47,13 +47,21 @@ static struct ble_gatt_svc_def espServices[ MAX_SERVICES + 1 ];
 static BTService_t * afrServices[ MAX_SERVICES ];
 static uint16_t serviceCnt = 0;
 static SemaphoreHandle_t xSem;
+static bool xSemLock;
 uint16_t gattOffset = 0;
 
 void prvGattGetSemaphore()
 {
+    xSemLock = 1;
     xSemaphoreTake( xSem, portMAX_DELAY );
 }
 
+void prvGattGiveSemaphore()
+{
+    xSemaphoreGive( xSem );
+    xSemLock = 0;
+}
+
```

Comment:
There could be a race condition here, if a request is already being processed, and a new request comes, it sets xSemLock = 1 and waits for the semaphore. Old request will post the semaphore, and resets xSemLock to 0. So it can cause prvValidRequest() to return false for the second request.

Reply:
Hi @ravibhagavandas 

> if a request is already being processed, and a new request comes, it sets xSemLock = 1 and waits for the semaphore. Old request will post the semaphore, and resets xSemLock to 0

This is not possible here because nimble host is running in a single thread. The same thread takes the semaphore (gets blocked till required content is received from AFR thread), and gives the semaphore when it is done using it (it has received required content). So there is no possibility of "_new request coming, when old request is being processed_ ". 
CC: @dhrishi Please correct me if you find anything wrong in the explanation.